### PR TITLE
Add `hearth-daemon` crate

### PR DESCRIPTION
**Pull request by marceline-cramer**
Paving the way for capability-based IPC. I migrated the stubs of the old IPC code into a new `hearth-daemon` crate and plugin. I also made it so that `hearth-core` finishes building plugins in the reverse order that they were added, so that we can say for sure that `hearth-daemon` adds its hook to `InitPlugin` before `InitPlugin` finalizes.

I copied the `Connection` struct from `hearth-network` into `hearth-ipc` because once we have WebSockets, we're not going to need to manually encode datagrams since WebSockets has its own datagram format built-in. IPC will need that datagram encoding for UDS (and potentially TCP for Windows IPC), so I just copied it there. When we add WebSockets support we can remove `hearth-network` entirely since it only includes `Connection` plus PAKE and encryption that will be replaced with TLS.

I also attempted to make `hearth-ipc` a little more transport-agnostic (for when we add some kind of Windows IPC in addition to UDS) by making the `connect()` function return a datagram `Connection` instead of a bare `UnixStream`.

Finally, `hearth-daemon` begins an actual cap exchange connection over the IPC transport. If the init system sends a registry to the init hook, then the IPC daemon will start. Theoretically, we should be able to do process interaction over IPC now, although this hasn't been tested yet. @airidaceae this means that we're ready to start working on `hearth-ctl` again.

- core: reverse-ordered plugin finish()
- Add initial hearth-daemon crate
- ipc: remove daemon-only code
- server: use hearth-daemon for IPC
- client: use hearth-daemon for IPC
- ipc: copy Connection from network crate
- daemon: begin cap exchange on accept
- daemon: send IPC client root cap

### CODE REVIEW

File: crates/hearth-daemon/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+        });
+    }
+}
+
+impl DaemonPlugin {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    pub fn on_accept(
```
[the remark this answers is not in the document]
Comment (airidaceae) [reply crates/hearth-daemon/src/lib.rs:157]: Not sure on this after reading it a bit more. Up to you.

File: crates/hearth-daemon/src/lib.rs
Comment (airidaceae) [inline crates/hearth-daemon/src/lib.rs:1]: Overall looks really good. All of this is pretty legible code. Excited to use it in `hearth-ctl` :purple_heart:

File: crates/hearth-ipc/src/lib.rs
Comment (airidaceae) [inline crates/hearth-ipc/src/lib.rs:1]: struggling to read this code but I think it may just be sleepiness. Going to give it another try tomorrow.

Comment (airidaceae) [reply crates/hearth-ipc/src/lib.rs:1]: ok nevermind i pulled up `hearth-deamon` and `hearth-ipc` up side-by-side and it makes sense now

**Review by airidaceae — CHANGES_REQUESTED** [review]
This all looks great. Pretty clean code and I made #162 for the only big issue I had with legibility. Id like the stuff I mentioned in this review to be considered before merging but its not all that much.

File: crates/hearth-daemon/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+                let msg = "Failed to find a socket path";
+                tracing::error!(msg);
+                return Err(Error::new(kind, msg));
+            }
+        };
+
+        match UnixStream::connect(&sock_path).await {
+            Ok(_) => {
+                tracing::warn!(
+                    "Socket is already in use. Another instance of Hearth may be running."
```
[the remark this answers is not in the document]
Comment (marceline-cramer) [reply crates/hearth-daemon/src/lib.rs:82]: Would you be alright if I removed the warning log and just print that message when `DaemonPlugin` panics on `unwrap()`? It's in a dedicated Tokio thread so it won't affect the rest of the runtime, it's just up to ergonomics of a log warning versus a panic message.

[the remark this answers is not in the document]
Comment (airidaceae) [reply crates/hearth-daemon/src/lib.rs:82]: I think that would work well for now. In the future id like to exit gracefully and map panics to an error-level log message.

**Review by airidaceae — APPROVED** [review]

**Pull request merged**